debezium-core/src/main/java/io/debezium/schema/TopicSelector.java

```diff
@@ -82,4 +88,77 @@ public String getHeartbeatTopic() {
     public interface DataCollectionTopicNamer<I extends DataCollectionId> {
         String topicNameFor(I id, String prefix, String delimiter);
     }
+
+    /**
+     * A topic namer that replaces any characters invalid in a topic name with {@code _}.
+     */
+    private static class TopicNameSanitizer<I extends DataCollectionId> implements DataCollectionTopicNamer<I> {
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(TopicNameSanitizer.class);
+
+        private static final String REPLACEMENT_CHAR = "_";
+
+        private final DataCollectionTopicNamer<I> delegate;
+
+        public TopicNameSanitizer(DataCollectionTopicNamer<I> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public String topicNameFor(I id, String prefix, String delimiter) {
+            String topicName = delegate.topicNameFor(id, prefix, delimiter);
+
+            StringBuilder sanitizedNameBuilder = new StringBuilder(topicName.length());
+            boolean changed = false;
+
+            for(int i = 0; i < topicName.length(); i++) {
+                char c = topicName.charAt(i);
+                if (isValidTopicNameCharacter(c)) {
+                    sanitizedNameBuilder.append(c);
+                }
+                else {
+                    sanitizedNameBuilder.append(REPLACEMENT_CHAR);
+                    changed = true;
+                }
+            }
+
+            if (changed) {
+                String sanitizedName = sanitizedNameBuilder.toString();
+                LOGGER.warn("Topic '{}' name isn't a valid topic name, replacing it with '{}'.", topicName, sanitizedName);
+
+                return sanitizedName;
+            }
+            else {
+                return topicName;
+            }
+        }
+
+        /**
+         * Whether the given character is a legal character of a Kafka topic name. Legal characters are
+         * {@code [a-zA-Z0-9._-]}.
+         *
+         * @see https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/internals/Topic.java
+         */
+        private boolean isValidTopicNameCharacter(char c) {
+            return c == '.' || c == '_' || c == '-' || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z')  || (c >= '0' && c <= '9');
+        }
+    }
+
+    /**
+     * A topic namer that caches names it has obtained from a delegate.
+     */
+    private static class TopicNameCache <I extends DataCollectionId> implements DataCollectionTopicNamer<I> {
+
+        private final ConcurrentMap<I, String> topicNames = new ConcurrentHashMap<>();
```

Comment:
This map is unbounded, I tend to think though it's ok. There won't be millions of tracked tables in a single connector, so it should be fine. WDYT, @jpechane?

Reply:
@gunnarmorling I think you are overoptimistic in this assumption - mind some of the installation we've already met. If we want to do the caching maybe it is time to introduce a proper LRU cache for this use case and maybe even others.

Reply:
Hum, hum. Ok. I'll think about it.

Reply:
Pushed one more commit to use a simple LRU cache here (based on `LinkedHashMap`).